### ISSUE TITLE
Number input breaks when letter "e" is entered

### ISSUE THREAD

**Issue by vahnag** [facebook/react#13499]
**Do you want to request a *feature* or report a *bug*?**
bug

**What is the current behavior?**
input[type=number] only allows entering numbers and letter "e" in the input. Native "input" input event is called for both numbers and the letter "e". With React the `onChange` event is only called for numbers. There's no way to catch "e" with `onChange`.

The "e" is even being filled when the input is controlled. The only way I can think of to work around this bug right now is to use `onKeyDown` and `preventDefault` when "e" or "E" is pressed.

**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem. Your bug will get fixed much faster if we can run your code and it doesn't have dependencies other than React. Paste the link to your JSFiddle (https://jsfiddle.net/Luktwrdm/) or CodeSandbox (https://codesandbox.io/s/new) example below:**
https://codesandbox.io/s/ov3ql3ljwz

**What is the expected behavior?**
It should pass anything that is being filled into the input to the onChange handler and should not break controlled component.

**Which versions of React, and which browser / OS are affected by this issue? Did this work in previous versions of React?**
React: 16.4.2
Chrome: 68.0.3440.106
Windows 10

**Comment by nhunzaker** [comment]
@Yurickh is spot on. Thanks! 

If I type the characters `1e2` I see three change events:

1. "1"
2. "" (because `1e` is not a valid number)
3. `"1e2"

I believe everything is behavior as intended. However please let me know if there's a case we've missed. Also this is a super confusing part of inputs, so I'm happy to continue the conversation if anything else needs clarification (or I've missed something).

**Comment by Ende93** [comment]
`onChange` get called, you can just see the console log, but when you input `1e`,  `e.target.value` is '', and `setState` get ''(empty string), so `e.target.value` is the same to setState, so React will not see the diff, and input will not change.

**Comment by nhunzaker** [comment]
That's correct. Numbers that start with `e` are invalid, so Chrome reports an empty string as `input.value`. As more numbers are added, the value reports the same (an empty string), so no change occurs.

**Comment by mgol** [comment]
@nhunzaker the problem here is that React's `onChange` doesn't allow to distinguish between empty inputs of type number and non-empty invalid ones. This is often needed, e.g. to implement Material Design-style floating labels that descend into the input when it's empty; see e.g. https://github.com/mui-org/material-ui/issues/12764. I assume this is because in both cases `value` is an empty string.

Modifying your steps, the scenario is as follows:

1. Enter "2" into the input - the callback fires.
2. Remove "2", making the input empty - the callback fires.
3. Enter "e" - the callback doesn't fire.

To implement Material UI inputs you need to check input's `validity.badInput` field. But you won't have a chance to actually see it.

Could `onChange` be fired not only if the input value changes but also if its `validity.badInput` field changes? It seems that if anything else is not valid with the input (e.g. the value is below `min` or `step` is not respected) the `value` field is not zeroed out.

Note that IE doesn't support `validity.badInput` but IE also doesn't always clear `value` for non-number values in a number input. For example, according to:
https://github.com/angular/angular.js/issues/10981#issuecomment-73231004
`0a` is not cleared in IE but `a0` is. This problem might not be solvable, though.

**Comment by gaearon** [comment]
@mgol Would #9657 resolve this?

**Comment by mgol** [comment]
@gaearon If I understand it correctly, this line:
> `onInput` (...) It is allowed to over-fire many events even if nothing changed.

suggests it should solve the issue as the browser fires the `input` event for each change, even if it doesn't change the `value` property because, for example, you type `e` into an empty `<input type="number">`.

I don't know what's the difference between the polyfilled and non-polyfilled `onInput`, is that relevant here? This particular issue affects both controlled and uncontrolled inputs.

**Comment by gaearon** [comment]
@mgol Is this issue about controlled or uncontrolled component? Or both?

**Comment by mgol** [comment]
Both.

**Comment by gaearon** [comment]
Tentatively reopening for future consideration.

**Comment by mgol** [comment]
This is still an issue.
